text/0116-sentry-semantic-conventions.md

```diff
@@ -0,0 +1,159 @@
+
+## Attributes Schema
+
+Attribute keys should be unique and well-known, and should not be used for multiple purposes. In OpenTelemetry the attribute values cannot be dictionaries, only primitives and arrays of primitives. We will try to follow this convention as well, but we will allow dictionaries as attribute values for better backwards compatibility with existing Sentry fields.
+
+```ts
+export interface Attributes {
+  [attributeKey: string]: AttributeValue;
+}
+
+/**
+ * Attribute values may be any non-nullish primitive value except an object.
+ *
+ * null or undefined attribute values are invalid and will result in undefined behavior.
+ */
+export type AttributeValue =
+  | string
+  | number
+  | boolean
+  | Array<string>
+  | Array<number>
+  | Array<boolean>
+  | Record<string, AttributeValue>;
+```
+
+```ts
+{
+  "http.method": "POST",
+  "http.request_content_length_uncompressed": 380,
+  "http.status_code": 200,
+  "http.status_text": OK,
+  "http.target": "/api/checkout",
+}
+```
+
+Whenever possible the attributes key/value meaning should be based on [OpenTelemetry's semantic conventions](https://opentelemetry.io/docs/concepts/semantic-conventions/). For example, the `http.request.method` attribute should have the same meaning as OpenTelemetry's `http.request.method` attribute. If a value does not exist in OpenTelemetry's semantic conventions, we should also aim to get it upstreamed into OpenTelemetry.
+
+The reason for having an independent set of semantic conventions is that we have some attributes that are not covered by OpenTelemetry's semantic conventions, and we don't want to be blocked on upstreaming them. There are also some Sentry specific attributes that will never match OpenTelemetry's semantic conventions.
+
+For sentry specific fields like `release` and `environment`, we are proposing setting them under the `sentry.X` namespace for the key. For example, `release` would be defined under the `sentry.release` attributes key.
+
+## Attributes Schema Versioning and Implementation
+
+The attributes schema will be versioned and published as a JSON Schema, which will be the representation of the well-known attributes that Sentry ingestion and product relies on.
+
+The JSON schema will also be codegened into packages for JavaScript, Python, and Rust, so that the schema can be used in Relay, the frontend and the backend. There has to be care put in to make sure that the codegened packages are compatible with the JSON schema, and that the JSON schema is backwards compatible, but we can introduce tests to make sure that this is the case.
+
+By having a versionig/packaging structure like this, it also makes it much easier for the SDKs to identify what parts of the product are using the data they are sending.
+
+## Mapping Sentry Specific Fields to Attributes
```

Comment:
So some things that are missing here, at least things I needed to manually handle in the POTEL experiments, are:

* span origin (`sentry.origin` ?)
* span op (`sentry.op` ?)
* span source (`sentry.source` ?)
* sample rate (`sentry.sample_rate` ?)

we also kind of have a parent sampled flag, but this may also just remain an internal thing I guess.

Reply:
Yeah I'm not sure this will ever get to an exhaustive state, more just important examples, but I think we need a good analog for trace context. Will add that.

Reply:
yeah, these were just the things I noticed we need to kind of maintain feature parity - esp. origin, op and source I believe? sample rate is maybe less important I guess.